### PR TITLE
Transplant easily generalized tests to VectorImplTestHelper from Test_DataVector

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Move tests which are easy to reformulate as general template tests from `Test_DataVector.cpp` to `VectorImplTestHelper.hpp`, and make use of the new test utilities instead. Also, the tests have been modernized to use random values and variable names and structure of the tests have been adjusted for clarity.
This is the 6th commit from #935.

### Depends on:
- [x] #1225 

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 1]
```diff
@@ -545,55 +401,15 @@ void test_datavector_math() {
...
+  INFO("test construct and assign");
+  TestHelpers::VectorImpl::vector_test_construct_and_assign<DataVector,
+                                                            double>();
+  INFO("test serialize and deserialize");
+  TestHelpers::VectorImpl::vector_test_serialize<DataVector, double>();
+  INFO("test set_data_ref functionality");
+  TestHelpers::VectorImpl::vector_test_ref<DataVector, double>();
+  INFO("test math after move");
+  TestHelpers::VectorImpl::vector_test_math_after_move<DataVector, double>();
+  SECTION("test DataVector math operations") { test_datavector_math(); }
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp]: Is this unintentionally a section?

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 2]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
...
+  UniformCustomDistribution<size_t> sdist{2, 20};
+
+  const size_t size = sdist(gen);
+  VectorType original_vector = make_with_random_values<VectorType>(
+      make_not_null(&gen), make_not_null(&dist), VectorType{size});
+
+  {
+    INFO("Check construction, copy, move, and ownership of reference vectors")
+    VectorType ref_vector;
+    ref_vector.set_data_ref(make_not_null(&original_vector));
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: don't need `make_not_null` here

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 3]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
...
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<typename tt::get_fundamental_type_t<ValType>> dist{
+      low, high};
+  UniformCustomDistribution<size_t> sdist{2, 20};
+
+  const size_t size = sdist(gen);
+  VectorType generated_vector = make_with_random_values<VectorType>(
+      make_not_null(&gen), make_not_null(&dist), VectorType{size});
+  VectorType ref_generated_vector;
+  ref_generated_vector.set_data_ref(make_not_null(&generated_vector));
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: I don't think you need the `make_not_null` anymore

**Review by nilsdeppe — COMMENTED** [review]
A couple minor things that were missed in the "fixup". I'm happy for you to squash. @kidder could you please look at this when you get a chance? Thanks!

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 4]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
       tup, std::make_index_sequence<N>(),
       std::make_index_sequence<sizeof...(ValueTypes) - N - 1>());
 }
+
+/// \ingroup TestingFrameworkGroup
+/// \brief test construction and assignment of a `VectorType` with a `ValType`
```
Comment (kidder) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: ValType -> ValueType

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 5]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
...
+  CHECK(serialized_vector_ref == vector_test);
+  CHECK(serialized_vector_ref.is_owning());
+  CHECK(serialized_vector_ref.data() != vector_ref.data());
+  CHECK_FALSE(vector_ref.is_owning());
+}
+
+/// \ingroup TestingFrameworkGroup
+/// \brief test the construction and move of a reference `VectorType`
+/// constructed with a `ValType`
+template <typename VectorType, typename ValType>
```
Comment (kidder) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: ValType -> ValueType

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 6]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
...
+/// function argument:
+/// - `RefSizeErrorTestKind::Copy`: Checks that copy-assigning to a non-owning
+/// `VectorType` from a `VectorType` with the wrong size generates an error.
+/// - `RefSizeErrorTestKind::ExpressionAssign`: Checks that assigning to a
+/// non-owning `VectorType` from an expression with `::ResultType` of
+/// `VectorType` with the wrong size generates an error
+/// - `RefSizeErrorTestKind::Move`: Checks that move-assigning to a non-owning
+/// `VectorType` from a `VectorType` with the wrong size generates an error.
+template <typename VectorType,
+          typename ValType = typename VectorType::ElementType>
```
Comment (kidder) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: ValueType

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 7]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
...
+  }
+  if(test_kind == RefSizeErrorTestKind::Move) {
+    ref_generated_vector = std::move(larger_generated_vector);
+  }
+}
+
+/// \ingroup TestingFrameworkGroup
+/// \brief tests a small sample of math functions after a move of a
+/// `VectorType` initialized with `ValType`
+template <typename VectorType, typename ValType>
```
Comment (kidder) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: ValueType

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — COMMENTED** [review]
I am happy for you to squash if @nilsdeppe is

**Comment by nilsdeppe** [comment]
I'm fine with a squash

**Comment by moxcodes** [comment]
squashed.
Cheers!

**Review by kidder — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]

**Comment by moxcodes** [comment]
This'll need a rebase following #1225, due to the template fix there

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 8]
```diff
@@ -160,5 +163,347 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
...
+
+enum RefSizeErrorTestKind { Copy, ExpressionAssign, Move };
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Test that assigning to a non-owning `VectorType` of the wrong size
+/// appropriately generates an error.
+///
+/// \details a calling function should be an `ASSERTION_TEST()` and check for
+/// the string "Must copy into same size".
+// Three types of tests are provided and one must be provided as the first
```
Comment (wthrowe) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: Missing `/`.

**Review by wthrowe — COMMENTED** [review]
Also note the Travis Doxy failure.

**Comment by nilsdeppe** [comment]
@moxcodes this can be rebased now :)

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 9]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
       tup, std::make_index_sequence<N>(),
       std::make_index_sequence<sizeof...(ValueTypes) - N - 1>());
 }
+
+/// \ingroup TestingFrameworkGroup
+/// \brief test construction and assignment of a `VectorType` with a `ValType`
+template <typename VectorType, typename ValType>
```
Comment (fmahebert) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: Should these `ValType`s go back to `ValueType`s?

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp [hunk 10]
```diff
@@ -160,5 +163,349 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
...
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<typename tt::get_fundamental_type_t<ValType>> dist{
+      low, high};
+  UniformCustomDistribution<size_t> sdist{2, 20};
+
+  const size_t size = sdist(gen);
+
+  const VectorType size_constructed{size};
+  CHECK(size_constructed.size() == size);
+  const auto generated_val1 = make_with_random_values<ValType>(
```
Comment (fmahebert) [inline tests/Unit/DataStructures/VectorImplTestHelper.hpp]: I have a mild preference for the `x_val` variable names to become `x_value` to match the use of `ValueType` as the template parameter. But feel free to veto.

**Review by fmahebert — COMMENTED** [review]
I really like the more descriptive variable names you are using now 👍. Here's two small naming things I found but that could be ignored.

File: tests/Unit/DataStructures/VectorImplTestHelper.hpp (hunk 9 above)
Comment (moxcodes) [reply tests/Unit/DataStructures/VectorImplTestHelper.hpp]: oh no! I have done git so very bad.

**Comment by moxcodes** [comment]
rebased, now with the actual edits in place. (thanks @fmahebert for noticing that)

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**